utils/filterDogBreeds.js

```diff
@@ -0,0 +1,108 @@
+// got some help from chatGPT to optimize my functions, check older commits for the process
+// it's still a lot of code, that's why I added some comments to explain the single functions
+
+export function filterDogBreeds(filter, dogData, setShowBreeds) {
+  let filteredBreeds = [];
+
+  if (filter.temperament !== "all" && filter.size !== "all") {
+    filteredBreeds = dogData.filter((breed) => handleAllFilters(breed, filter));
+  }
+
+  // check if the group filter is set
+  else if (filter.group !== "all") {
+    // check if the group and temperament filter is set
+    if (filter.temperament !== "all") {
+      filteredBreeds = dogData.filter((breed) =>
+        handleGroupAndTemperament(breed, filter)
+      );
+    }
+    // check if the group and size filter is set
+    else if (filter.size !== "all") {
+      filteredBreeds = dogData.filter((breed) =>
+        handleGroupAndSize(breed, filter)
+      );
+    }
+    // check if ONLY the group filter is set
+    else {
+      filteredBreeds = dogData.filter(
+        (breed) => breed.breed_group === filter.group
+      );
+    }
+  }
+  // check if the temperament filter is set
+  else if (filter.temperament !== "all") {
+    // check if the temperament and size filter is set
+    if (filter.size !== "all") {
+      filteredBreeds = dogData.filter((breed) =>
+        handleTemperamentAndSize(breed, filter)
+      );
+    }
+    // check if ONLY the temperament filter is set
+    else {
+      filteredBreeds = dogData.filter(
+        (breed) =>
+          breed.temperament && breed.temperament.includes(filter.temperament)
+      );
+    }
+  }
+  // check if only the size filter is set
+  else if (filter.size !== "all") {
+    filteredBreeds = dogData.filter(
+      (breed) =>
+        (filter.size === "small" && getHeight(breed) < 4) ||
+        (filter.size === "medium" &&
+          getHeight(breed) < 6 &&
+          getHeight(breed) >= 4) ||
+        (filter.size === "large" && getHeight(breed) >= 6)
+    );
+  } else {
+    filteredBreeds = dogData;
+  }
+
+  setShowBreeds(filteredBreeds);
```

Comment:
I think you wrote this function more complicated than it needed to be. you can chain 3 .filter methods after one another and you will get a similar result:
```js
const filteredByTemperatment = filter.temperament === 'all' ? dogData : dogData.filter(dog => ...)
const filteredByBreed = filter.breed === 'all' ? filteredByTemerament : filteredByTemerament.filter(dog => ...)
const filteredBySize = filter.size === 'all' ? filteredByBreed : filteredByBreed.filter(dog => ...)
return filteredBySize
```